### PR TITLE
Optimize mask memory for flash attention

### PR DESCRIPTION
**Pull request by yanbing-j**
The PR optimizes the mask memory for flash attention. Instead of directly converting the whole mask to fp32, we do the conversion block-wisely. This can decrease the peak memory usage (we test in https://huggingface.co/microsoft/Phi-3-mini-128k-instruct, peak memory usage reduces ~50%) and have some performance improvements as well.

### Performance result
single socket in Intel (R) Xeon (R) CPU Max 9480
batch_size = 12, q_seq_len = 1030, kv_seq_len = 1179, n_head = 3, head_dim = 33, mask_dim = 4, bool_mask = 0
  | Forward speedup | Backward speedup
-- | -- | --
float64 | 0.82% | 3.76%
float32 | 2.2% | 3.9%
bfloat16 | 16.15% | 7.56%

**segment-anything-fast**
Follow https://github.com/pytorch-labs/segment-anything-fast/tree/main/experiments
Single socket in Intel (R) Xeon (R) CPU Max 9480
Dtype: bfloat16, models: vit_b and vit_h, test in `SDPA` and `Triton` commit https://github.com/pytorch-labs/segment-anything-fast/blob/main/experiments/run_experiments.py#L199-L200, select the time of 20th iteration.
  | vit_b |   | vit_h |  
-- | -- | -- | -- | --
  | attn_mask w/o   block-wise | attn_mask w/   block-wise | attn_mask w/o   block-wise | attn_mask w/   block-wise
SDPA| 10.95s/it | 6.59s/it | 19.93s/it | 12.33s/it
Triton | 10.66s/it | 7.12s/it | 19.87s/it | 12.26s/it

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by Valentine233 — APPROVED** [review]
LGTM.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 1]
```diff
@@ -21,6 +21,53 @@ namespace at::native {
...
+  for (long i = vec_size * (size / vec_size); i < size; i++) {
+    auto tmp0 = a[i];
+    auto tmp1 = (T1) b[i];
+    out[i] = tmp0 * val + tmp1;
+  }
+}
+
+// out = val * a + b
+template <typename T1>
+inline void _scale_attn_mask_fusion_kernel(
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:50]: NIT: Maybe you don't have to add this one and can reuse the function above. You can make `vec::convert` on the same data type as a nop (by adding a specialization to `vec_convert.h`).

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 1 above)
Comment (yanbing-j) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp]: Now, vec::convert can support with the same data type as a noop.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 2]
```diff
@@ -21,6 +21,62 @@ namespace at::native {
...
+    T1& val) {
+  auto vec_size1 = at::vec::Vectorized<T1>::size();
+  auto vec_size2 = at::vec::Vectorized<T2>::size();
+  auto vec_scale = at::vec::Vectorized<T1>(val);
+  int64_t i = 0;
+  for (; i < size - (size % vec_size2); i += vec_size2) {
+    auto a0 = at::vec::Vectorized<T1>::loadu(a + i);
+    auto a1 = at::vec::Vectorized<T1>::loadu(a + i + vec_size1);
+    auto b_tmp = at::vec::Vectorized<T2>::loadu(b + i);
+    auto [b0, b1] = convert_to_float<T2>(b_tmp);
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:41]: You can probably unify the two functions into one the vec convert too, with vectorizedn.

Comment (yanbing-j) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp]: Done. Please take a look.

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**